### PR TITLE
add a guide on ahead-of-time compilation

### PR DESCRIPTION
**Pull request by sayakpaul**

### CODE REVIEW

File: inference/aot_serialization.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+def aot_compile(name, fn, **sample_kwargs):
+    path = f"{name}.so"
+    print(f"{path=}")
+    options = {
+        "aot_inductor.output_path": path,
+        "max_autotune": True,
+        "triton.cudagraphs": True,
+    }
+
+    torch._export.aot_compile(
```
Comment (jerryzh168) [inline inference/aot_serialization.md:146]: what's the plan for using the new api? `torch._inductor.aoti_compile_and_package`, do you want to update this after Bin fixed the model size issue?

File: inference/aot_serialization.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+image = pipeline(
+    "cute dog", guidance_scale=3.5, max_sequence_length=512, num_inference_steps=50
+).images[0]
+image.save("aot_compiled.png")
+```
+
+Note that running the above requires the following patch:
+
+```diff
```
Comment (jerryzh168) [inline inference/aot_serialization.md:203]: why is this needed? is this something specific to flux or some bug in aot_compile/load?

File: inference/aot_serialization.md (hunk 1 above)
Comment (sayakpaul) [reply inference/aot_serialization.md:146]: Yeah exactly. The gist that tackles the quantization + AoT part still uses `torch._inductor.aoti_compile_and_package`, though.

File: inference/aot_serialization.md (hunk 2 above)
Comment (sayakpaul) [reply inference/aot_serialization.md:203]: Specific the AoT compiled function.

**Comment by sayakpaul** [comment]
@jerryzh168 okay with you with the rest of the PR?

File: inference/aot_serialization.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    )
+    return f"{(t0.blocked_autorange().mean):.3f}"
+
+def load_model():
+    model = FluxTransformer2DModel.from_pretrained(
+        "black-forest-labs/FLUX.1-dev", subfolder="transformer", torch_dtype=torch.bfloat16
+    ).to("cuda")
+    return model
+
+def aot_compile(name, fn, **sample_kwargs):
```
Comment (jerryzh168) [inline inference/aot_serialization.md:137]: wondering if you want to make this a bit more general to allow `sample_args` as well in case people want to copy paste

File: inference/aot_serialization.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        options=options,
+        disable_constraint_solver=True,
+    )
+    return path 
+
+def aot_load(path):
+    return torch._export.aot_load(path, "cuda")
+
+@torch.no_grad()
+def f(model, **kwargs):
```
Comment (jerryzh168) [inline inference/aot_serialization.md:159]: and this one as well

File: inference/aot_serialization.md (hunk 3 above)
Comment (sayakpaul) [reply inference/aot_serialization.md:137]: Could you put that as a suggestion to better denote what you mean?

File: inference/aot_serialization.md (hunk 2 above)
Comment (jerryzh168) [reply inference/aot_serialization.md:203]: I'm trying to understand how people will use this.

Is this tutorial going to be used by other model code as well? if so, how do they know how to make similar changes to their model

**Review by jerryzh168 — APPROVED** [review]
yeah the rest looks good to me

File: inference/aot_serialization.md (hunk 2 above)
Comment (sayakpaul) [reply inference/aot_serialization.md:203]: They will apply similar changes to their pipeline code to make it compatible.

Comment (jerryzh168) [reply inference/aot_serialization.md:203]: I see, that sounds good, I think it's fine

**Review by a-r-r-o-w — APPROVED** [review]
awesome!

File: inference/aot_serialization.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply inference/aot_serialization.md:137]: And where is `sample_args` used down further?

File: inference/aot_serialization.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply inference/aot_serialization.md:159]: Same here as well.

File: inference/aot_serialization.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    print(f"{path=}")
+    options = {
+        "aot_inductor.output_path": path,
+        "max_autotune": True,
+        "triton.cudagraphs": True,
+    }
+
+    torch._export.aot_compile(
+        fn,
+        (),
```
Comment (jerryzh168) [inline inference/aot_serialization.md:148]: cc @sayakpaul sorry just saw the comment, `sample_args` should be used here

Comment (sayakpaul) [reply inference/aot_serialization.md:148]: Would you be able to open a PR by updating this method to use
`torch._inductor.aoti_compile_and_package()`?
